File: internal/device/phone_mode.go
package device

import (
	"strings"

	"github.com/iniwex5/vowifi-go/runtimehost/carrier"
	"github.com/yibaiba/hideck/internal/cardpolicy"
	"github.com/yibaiba/hideck/internal/config"
)

const (
	PhoneModeWiFi     = "wifi"
	PhoneModeCellular = "cellular"
	PhoneModeVoLTE    = "volte"
)

func IsNativeVoLTEMode(mode string) bool {
	return strings.TrimSpace(mode) == PhoneModeVoLTE
}

func PhoneModeCampsOnCell(mode string) bool {
	switch strings.TrimSpace(mode) {
	case PhoneModeCellular, PhoneModeVoLTE:
		return true
	default:
		return false
	}
}

// PhoneServiceEnabled reports the persisted "phone on" bit.
// The card-policy column remains VoWiFiEnabled; native VoLTE and
// cellular software IMS reuse it as the phone-service switch.
func PhoneServiceEnabled(cfg config.DeviceConfig) bool {
	return cfg.VoWiFiEnabled
}

func cellularAlwaysData(cfg config.DeviceConfig) bool {
	return strings.TrimSpace(cfg.PhoneMode) == PhoneModeCellular && strings.TrimSpace(cfg.DataStrategy) == "always"
}

// WorkerSoftwareIMSBlocked reports China MCC 460/461, which has no ePDG
// software IMS path in this stack. Phone service must use native VoLTE.
func WorkerSoftwareIMSBlocked(w *Worker) bool {
	if w == nil {
		return false
	}
	mcc := strings.TrimSpace(w.state.Identity.NativeMCC)
	if carrier.IsVoWiFiBlockedMCC(mcc) {
		return true
	}
	imsi := strings.TrimSpace(w.state.Identity.IMSI)
	if len(imsi) >= 3 && carrier.IsVoWiFiBlockedMCC(imsi[:3]) {
		return true
	}
	return false
}

func forceSoftwareIMSBlockedToVoLTE(w *Worker, pol cardpolicy.Policy) {
	if w == nil || !PhoneServiceEnabled(w.Config) || !WorkerSoftwareIMSBlocked(w) {
		return
	}
	w.Config.PhoneMode = PhoneModeVoLTE
	if !pol.AirplaneEnabled {
		w.Config.AirplaneEnabled = false
	}
}

// applyPhoneRadioPolicy applies RF side-effects of turning phone service on.
// Camped modes (cellular, volte) keep RF on; WiFi calling takes the radio.
func applyPhoneRadioPolicy(cfg *config.DeviceConfig) {
	if cfg == nil || !PhoneServiceEnabled(*cfg) {
		return
	}
	if PhoneModeCampsOnCell(cfg.PhoneMode) {
		cfg.AirplaneEnabled = false
		if cellularAlwaysData(*cfg) {
			cfg.NetworkEnabled = true
		}
		return
	}
	cfg.AirplaneEnabled = true
	cfg.NetworkEnabled = false
}

File: web/src/utils/phoneMode.ts
/** 通话方式：wifi 软件 VoWiFi；cellular 软件 IMS 走蜂窝数据；volte 模组原生 IMS。 */
export function normalizePhoneMode(mode?: string | null): string {
  const value = (mode ?? 'wifi').trim()
  return value || 'wifi'
}

export function isNativeVoLTEMode(mode?: string | null): boolean {
  return normalizePhoneMode(mode) === 'volte'
}

export function phoneModeCampsOnCell(mode?: string | null): boolean {
  const value = normalizePhoneMode(mode)
  return value === 'cellular' || value === 'volte'
}

export function isWifiCallingEnabled(mode?: string | null, phoneEnabled?: boolean): boolean {
  return !!phoneEnabled && !phoneModeCampsOnCell(mode)
}

/** Ut/XCAP 只挂在软件 IMS 上；原生 VoLTE 和未开 WiFi calling 的设备不展示。 */
export function deviceSupportsUt(device?: {
  vowifi_enabled?: boolean
  vowifi_active?: boolean
  phone_mode?: string
} | null): boolean {
  if (!device || isNativeVoLTEMode(device.phone_mode)) return false
  return device.vowifi_enabled === true || device.vowifi_active === true
}

export function softwareIMSBlocked(device?: { software_ims_blocked?: boolean } | null): boolean {
  return device?.software_ims_blocked === true
}

export function phoneModeLabel(mode?: string | null): string {
  const value = normalizePhoneMode(mode)
  if (value === 'volte') return 'VoLTE'
  if (value === 'cellular') return '蜂窝数据'
  return 'WiFi calling'
}
